FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/viewmodel/RefreshPrivateKeysFromEkmViewModel.kt

```diff
@@ -0,0 +1,233 @@
+        context = context,
+        ekmUrl = requireNotNull(activeAccount.clientConfiguration?.keyManagerUrl),
+        idToken = idToken
+      )
+
+      if (ekmPrivateResult.status != Result.Status.SUCCESS) {
+        return@withContext when (ekmPrivateResult.status) {
+          Result.Status.EXCEPTION -> {
+            Result.exception(
+              //Thrown network errors can be ignored but not other types of errors
+              GeneralUtil.preProcessException(
+                context = context,
+                causedException = requireNotNull(ekmPrivateResult.exception)
+              )
+            )
+          }
+
+          Result.Status.ERROR -> Result.exception(ApiException(ekmPrivateResult.data?.apiError))
+
+          else -> throw IllegalStateException(
+            "Unsupported status = ${ekmPrivateResult.status} at this step"
+          )
+        }
+      }
+
+      if (ekmPrivateResult.data?.privateKeys?.isEmpty() == true) {
+        return@withContext Result.success(true)
+      }
+
+      val pgpKeyDetailsList = mutableListOf<PgpKeyDetails>()
+      ekmPrivateResult.data?.privateKeys?.forEach { key ->
+        val parsedList = PgpKey.parsePrivateKeys(requireNotNull(key.decryptedPrivateKey))
+          .map {
+            it.copy(
+              passphraseType = KeyEntity.PassphraseType.RAM,
+              importSourceType = KeyImportDetails.SourceType.EKM,
+              tempPassphrase = null
+            )
+          }
+
+        if (parsedList.isEmpty()) {
+          //ask Tom should we do any things here
+          throw IllegalStateException(context.getString(R.string.could_not_parse_one_of_ekm_key))
+        } else {
+          //ask Tom should we check isFullyDecrypted
+
+          //check that all keys were fully decrypted when we fetched them.
+          // If any is encrypted at all, that's an unexpected error, we should throw an exception.
+          parsedList.forEach {
+            if (!it.isFullyDecrypted) {
+              throw IllegalStateException(
+                context.getString(
+                  R.string.found_not_fully_decrypted_key_ask_admin,
+                  it.fingerprint
+                )
+              )
+            }
+          }
+          pgpKeyDetailsList.addAll(parsedList)
+        }
+      }
+
+      handleParsedKeys(activeAccount, pgpKeyDetailsList)
+
+      Result.success(true)
+    }
+
+  private suspend fun handleParsedKeys(
+    activeAccount: AccountEntity,
+    fetchedPgpKeyDetailsList: MutableList<PgpKeyDetails>
+  ) = withContext(Dispatchers.IO) {
+    val context: Context = getApplication()
+    val keysStorage = KeysStorageImpl.getInstance(context)
+    val existingKeyEntities = keysStorage.getRawKeys()
+    val existingPgpKeyDetailsList = keysStorage.getPgpKeyDetailsList()
+    var passphrase: Passphrase? = null
+
+    for (fetchedPgpKeyDetails in fetchedPgpKeyDetailsList) {
+      if (!fetchedPgpKeyDetails.usableForEncryption) {
+        //ask Tom
```

Comment:
@tomholub Please pay attention to this step

Reply:
Agree. What about encryption things? Should I check that a key can be used for encryption? Maybe the administrator made an error and posted a key that can be used just for signing.

Reply:
Expired and revoked keys cannot be used for encryption. And yet we still want to update these keys because they were made expired or revoked for a reason.

Therefore, we treat unusable keys the same as usable here.